### DISCUSSION TITLE
Creating GrainId

### DISCUSSION THREAD

**Discussion by jbockle** [dotnet/orleans#7617] [Q&A] [answered]
I'm creating a library where I want the consumer to choose what the primary key type is - I see I can use GrainId for this, is there a better way of instiating one? I'm doing the following currently:

```csharp
public interface IMyLibraryGrain : IGrain
{
    Task Foo();
}

// and user can create a a GrainId using an extension method something like below..
public static GrainId CreateMyLibraryGrainId(this IGrainFactory grainFactory, Guid|long|string key)
{
    return grainFactory.GetGrain(typeof(IMyLibraryGrain), key).GetGrainId();
}
```

**Comment by ReubenBond** [comment, marked as answer]
I'm not entirely sure what you're after, but perhaps I can supply a bit of detail:

A `GrainId` consists of two components, `Type` and `Key`, both are essentially strings.
Orleans treats the `Key` component as opaque.
The `Type` components has some meaning because it maps to grain implementation classes and if those classes are generic then the `Type` component will contain the encoded generic parameters (eg, ```my-grain`2[[int],[string]]```).

`Orleans.Runtime.GrainIdKeyExtensions` is a static class with methods for creating `Guid` & `long` keys (encoded as strings)
You can use this to create `Guid`/`long` keys. For string keys, you can pass the string directly to `GrainId.Create(...)`. These keys have no special meaning to Orleans - they're just strings.

The trickier part is the grain type. If you know what the name of the grain type you need is (eg, because it's using the default convention-based grain type where `UserGrain` => `user`, or because you put a `[GrainType("user-profile")]` attribute on the grain *class*), you can pass that as a string. Otherwise, you have to do a 2-step dance:

* Take your `TInterface` and resolve the corresponding `GrainInterfaceType` (also a string: the encoded name of the interface)
* Take your `GrainInterfaceType` and resolve the default `GrainType` which implements it (you can override the default by putting `[DefaultGrainType("user-profile")]` on the grain *interface*). To do this, you can use `GrainInterfaceTypeToGrainTypeResolver`.

Both of these types can be injected. So, a `GrainId` creator class might look like this:

``` C#
public class GrainIdFactory
{
    private readonly GrainInterfaceTypeResolver _grainInterfaceTypeResolver;
    private readonly GrainInterfaceTypeToGrainTypeResolver _grainInterfaceTypeToGrainTypeResolver;

    public GrainIdFactory(
        GrainInterfaceTypeResolver interfaceTypeResolver,
        GrainInterfaceTypeToGrainTypeResolver interfaceToTypeResolver)
    {
        _grainInterfaceTypeResolver = interfaceTypeResolver;
        _grainInterfaceTypeToGrainTypeResolver = interfaceToTypeResolver;
    }

    public GrainId CreateGrainId<TInterface>(Guid grainKey) where TInterface : IAddressable
    {
        GrainType grainType = ResolveGrainType(typeof(TInterface));
        GrainId grainId = GrainId.Create(grainType, GrainIdKeyExtensions.CreateGuidKey(grainKey));
        return grainId;
    }

    public GrainId CreateGrainId<TInterface>(long grainKey) where TInterface : IAddressable
    {
        GrainType grainType = ResolveGrainType(typeof(TInterface));
        GrainId grainId = GrainId.Create(grainType, GrainIdKeyExtensions.CreateIntegerKey(grainKey));
        return grainId;
    }

    public GrainId CreateGrainId<TInterface>(string grainKey) where TInterface : IAddressable
    {
        GrainType grainType = ResolveGrainType(typeof(TInterface));
        GrainId grainId = GrainId.Create(grainType, grainKey);
        return grainId;
    }

    private GrainType ResolveGrainType(Type interfaceType)
    {
        GrainInterfaceType grainInterfaceType = _grainInterfaceTypeResolver.GetGrainInterfaceType(interfaceType);
        GrainType grainType = _grainInterfaceTypeToGrainTypeResolver.GetGrainType(grainInterfaceType);
        return grainType;
    }
}
```

One important thing to note: the mapping from `GrainInterfaceType` to `GrainType` is reliant on the silo being started, since it needs to build up that mapping. In the case of a client, it relies on the client being connected to the cluster, since it needs to fetch the mapping from the cluster (the type which does this is `IClusterManifestProvider`).

Hope that helps.

**Reply by jbockle** [reply]
thanks for the info
